### PR TITLE
Consolidate metadata and dataset loading using a hook and React context

### PR DESCRIPTION
**Pull request by aaronsn**
- Ensure that the metadata and datasets are each only loaded once, and the results are cached for later use
- Provide a central place to update dataset loading logic
- Fix a race condition with data previews where clicking a slow one followed by a fast one would result in the data switching when the request finishes.
- Add a basic Logger implementation that logs to console in debug mode. This is just for convenience for debugging, but depending on our approach to logging later on we can probably use it as an entry point for logging.

### CODE REVIEW

File: app/src/utils/DatasetTypes.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    return this.metadata.fields.map((field: Field) => ({
+      Header: field.name,
+      accessor: field.name,
+    }));
+  }
+}
+
+export type AllMetadata = Record<string, DatasetMetadata>;
+
+export type LoadStatus = "unloaded" | "loading" | "loaded" | "error";
```
Comment (kkatzen) [inline app/src/utils/DatasetTypes.ts:39]: Could this be an enum? https://www.typescriptlang.org/docs/handbook/enums.html
I think it's easier to make mistakes when relying on strings to match whereas with enums you'll have compiler problems if you don't use a valid value

File: app/src/utils/Logger.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+export type Severity = "INFO" | "WARNING" | "ERROR";
```
Comment (kkatzen) [inline app/src/utils/Logger.ts:1]: maybe an enum?  https://www.typescriptlang.org/docs/handbook/enums.html

File: app/src/utils/DatasetTypes.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+export interface DatasetStore {
+  readonly loadDataset: (id: string) => void;
+  readonly getDatasetLoadStatus: (id: string) => LoadStatus;
+  readonly metadataLoadStatus: LoadStatus;
+  readonly metadata: AllMetadata;
```
Comment (kkatzen) [inline app/src/utils/DatasetTypes.ts:45]: I'm not sure I understand why you're creating a type for AllMetadata but not Record<string, Dataset>
My instinct is to prefer having Record versions for both as you get more visibility into what the object is without having to look up AllMetadata definition

If you do feel there's some value in keeping the type, I would add a comment at definition explaining that the string is the dataset ID to be explicit

File: app/src/features/dataset_explorer/DatasetExplorer.tsx [hunk 4]
```diff
@@ -1,76 +1,61 @@
...
+import { DatasetStore } from "../../utils/DatasetTypes";
 
 function renderTableOrPlaceholder(
-  loadStatus: LoadStatus,
-  columns: any[],
-  data: any[]
+  datasetStore: DatasetStore,
+  datasetId: string
 ) {
+  const loadStatus = datasetStore.getDatasetLoadStatus(datasetId);
```
Comment (kkatzen) [inline app/src/features/dataset_explorer/DatasetExplorer.tsx:14]: can this variable go directly in the switch?

File: app/src/features/dataset_explorer/DatasetExplorer.tsx [hunk 5]
```diff
@@ -1,76 +1,61 @@
...
 import DatasetListing from "./DatasetListing";
 import styles from "./DatasetExplorer.module.scss";
 import ChevronRightIcon from "@material-ui/icons/ChevronRight";
-import { DatasetMetadata } from "../../utils/DatasetMetadata";
-
-type LoadStatus = "unloaded" | "loading" | "loaded";
+import { useDatasetStore } from "../../utils/useDatasets";
+import { DatasetStore } from "../../utils/DatasetTypes";
 
 function renderTableOrPlaceholder(
```
Comment (kkatzen) [inline app/src/features/dataset_explorer/DatasetExplorer.tsx:10]: nit: nlow that this page also has fields table, maybe we should update this to renderPreviewTableOrPlaceholder or even just function PreviewTable. Also, is it not technically its own component? Can it get DatasetStore directly instead of params?

File: app/src/utils/useDatasets.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+ * @param cacheManager The ResourceCacheManager that tracks that resource.
+ * @param loadFunction A function that loads the resource when called.
+ */
+async function loadResource<R>(
+  resourceId: string,
+  cacheManager: ResourceCacheManager<R>,
+  loadFunction: () => Promise<R>
+) {
+  try {
+    if (cacheManager.cache.statuses[resourceId]) {
```
Comment (kkatzen) [inline app/src/utils/useDatasets.ts:37]: As we talked in slack, I think relying on truthiness of a string feels brittle to me, especially since some of the valid options on the string is "unloaded" and "error". I worry someone may disrupt this functionality with future changes

File: app/src/utils/DatasetTypes.ts (hunk 1 above)
Comment (aaronsn) [reply app/src/utils/DatasetTypes.ts:39]: As far as typechecking goes, enums and type unions provide mostly equivalent compile-time safety. If you try to assign an arbitrary string to a variable of type `"unloaded" | "loading" | "loaded" | "error"` you'll get a compiler error. I did a bit of research on this and it seems like this is mostly a matter of preference. The main reason to use an enum would be if you need to iterate over the possible values, and if you prefer it for readability.

I have a preference for a type union for simple cases like this because you don't need to import the type every time to use it and because it doesn't generate weird transpiled JS which could make debugging difficult.

LMK if you have concerns with that.

https://stackoverflow.com/questions/40275832/typescript-has-unions-so-are-enums-redundant
https://medium.com/@maxheiber/alternatives-to-typescript-enums-50e4c16600b1

File: app/src/utils/Logger.ts (hunk 2 above)
Comment (aaronsn) [reply app/src/utils/Logger.ts:1]: See above

File: app/src/utils/DatasetTypes.ts (hunk 3 above)
Comment (aaronsn) [reply app/src/utils/DatasetTypes.ts:45]: I was using this data type all over the place and it was pretty verbose. Also, comparing to the Dataset type, it's more common to need a reference to the whole metadata map than it is for the datasets where you'd usually just care about a specific one - eg metadata is one API call, and the catalog expects the whole map. I also updated it to be a read-only type to prevent mutating it without react noticing, which made it more verbose.

As an example, these were what it looked like before and I found that kind of hard to parse:
```
useResourceCache<Readonly<Record<string, Dataset>>>()
Promise<Readonly<Record<string, Dataset>>>
```

I changed the name to MetadataMap and a comment for the structure to make it clearer

File: app/src/features/dataset_explorer/DatasetExplorer.tsx (hunk 4 above)
Comment (aaronsn) [reply app/src/features/dataset_explorer/DatasetExplorer.tsx:14]: Yup

File: app/src/features/dataset_explorer/DatasetExplorer.tsx (hunk 5 above)
Comment (aaronsn) [reply app/src/features/dataset_explorer/DatasetExplorer.tsx:10]: Good call on both accounts. I moved this to its own file. I also extracted the "select a dataset to view" so that the data preview can be more portable and used in another context

File: app/src/utils/useDatasets.ts (hunk 6 above)
Comment (aaronsn) [reply app/src/utils/useDatasets.ts:37]: I pulled this out into a helper function shouldLoadResource that returns true if the load status is undefined or "unloaded" or "error". We can adjust that logic later if needed

**Review by kkatzen — APPROVED** [review]
All sounds good to me :)
